### ISSUE TITLE
Xamarin.iOS System.ObjectDisposedException

### ISSUE THREAD

**Issue by craylward** [Azure/azure-iot-sdk-csharp#606]
### Description

Intermittent crash on Xamarin.iOS application using DeviceClient.RecieveAsync() .

Visual Studio 15.8.1
Xamarin.iOS 11.14.0.13
Microsoft.Azure.Devices.Client 1.17.1
TransportType.Amqp_WebSocket_Only

```
018-08-31 21:32:18.032 [ERROR] (2:Finalizer) [UnhandledExceptionReporter] : Reporting unhandled exception: TaskSchedulerOnUnobservedTaskException System.Exception: TaskSchedulerOnUnobservedTaskException ---> System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'FaultTolerantAmqpObject`1'.
  at Microsoft.Azure.Devices.Client.Singleton`1+<GetOrCreateAsync>d__12[TValue].MoveNext () [0x000fc] in <b97b5de576ca414c83d9ecd5bc8def92>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <c4da4bcb0a614f31bf9f25261a36b747>:0 
  at Microsoft.Azure.Devices.Client.IotHubConnection+<GetSessionAsync>d__21.MoveNext () [0x00109] in <b97b5de576ca414c83d9ecd5bc8def92>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <c4da4bcb0a614f31bf9f25261a36b747>:0 
  at Microsoft.Azure.Devices.Client.IotHubConnection+<CreateReceivingLinkAsync>d__20.MoveNext () [0x000a2] in <b97b5de576ca414c83d9ecd5bc8def92>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <c4da4bcb0a614f31bf9f25261a36b747>:0 
  at Microsoft.Azure.Devices.Client.Transport.AmqpTransportHandler+<CreateTwinReceivingLinkAsync>d__83.MoveNext () [0x000cd] in <b97b5de576ca414c83d9ecd5bc8def92>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <c4da4bcb0a614f31bf9f25261a36b747>:0 
  at Microsoft.Azure.Devices.Client.FaultTolerantAmqpObject`1+<OnCreateAsync>d__5[T].MoveNext () [0x0009f] in <b97b5de576ca414c83d9ecd5bc8def92>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <c4da4bcb0a614f31bf9f25261a36b747>:0 
  at Microsoft.Azure.Devices.Client.Singleton`1+<CreateValue>d__19[TValue].MoveNext () [0x0007e] in <b97b5de576ca414c83d9ecd5bc8def92>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <c4da4bcb0a614f31bf9f25261a36b747>:0 
  at Microsoft.Azure.Devices.Client.Singleton`1+<GetOrCreateAsync>d__12[TValue].MoveNext () [0x0008f] in <b97b5de576ca414c83d9ecd5bc8def92>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <c4da4bcb0a614f31bf9f25261a36b747>:0 
  at Microsoft.Azure.Devices.Client.Transport.AmqpTransportHandler+<GetTwinReceivingLinkAsync>d__82.MoveNext () [0x00093] in <b97b5de576ca414c83d9ecd5bc8def92>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00037] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in <c4da4bcb0a614f31bf9f25261a36b747>:0 
  at Microsoft.Azure.Devices.Client.Transport.AmqpTransportHandler+<EnableTwinReceivingLinkAsync>d__54.MoveNext () [0x00071] in <b97b5de576ca414c83d9ecd5bc8def92>:0 
--- End of stack trace from previous location where exception was thrown ---
  at Microsoft.Azure.Devices.Client.Transport.AmqpTransportHandler+<>c__DisplayClass49_0+<<EnableTwinPatchAsync>b__0>d.MoveNext () [0x00168] in <b97b5de576ca414c83d9ecd5bc8def92>:0 
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
  at PatientApp.iOS.AppDelegate.UnobservedTaskExceptionHandler (System.Object sender, System.Threading.Tasks.UnobservedTaskExceptionEventArgs args) [0x00012] in C:\NGQ_Workspace\Source_Git\pa_x\PatientApp\PatientApp.iOS\AppDelegate.cs:275 
  at System.Threading.Tasks.TaskScheduler.PublishUnobservedTaskException (System.Object sender, System.Threading.Tasks.UnobservedTaskExceptionEventArgs ueea) [0x00019] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/threading/Tasks/TaskScheduler.cs:529 
  at System.Threading.Tasks.TaskExceptionHolder.Finalize () [0x000e6] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.13/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/threading/Tasks/TaskExceptionHolder.cs:142 
```

**Comment by CIPop** [comment]
@craylward Please try with DeviceClient 1.18.0: we've made some changes to the WebSocket implementation for netstandard2.0.

Are you sure you've pasted the right stack? This seems to be identical to #605.

**Comment by CIPop** [comment]
This is a duplicate of #265. For .NET Desktop we've decided we won't be fixing this in v1.x: tasks are started and, if they Timeout they will be left running in the background until they either throw or complete. We need to implement proper cooperative cancellation (CancellationToken) instead of starting timers.

Closing as duplicate of #528 - while this appears to be the same issue as #265, Xamarin is impacted. @craylward - is it possible to implement the workaround I've presented here: https://github.com/Azure/azure-iot-sdk-csharp/issues/528#issuecomment-400899315 ?

**Comment by az-iot-builder-01** [comment]
@craylward, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
